src/limbo/experimental/model/ssgpr.hpp

```diff
@@ -169,8 +170,59 @@ namespace limbo {
                     _sigma_n = std::exp(p(0)); //p(0); //std::exp(p(0)) + 1e-6;
                 }
 
+            /// add sample and update the GP. This code uses an incremental implementation of the Cholesky
+            /// decomposition. It is therefore much faster than a call to compute()
+            void add_sample(const Eigen::VectorXd& sample, const Eigen::VectorXd& observation)
+            {
+                if (_samples.empty()) {
+                    if (_dim_in != sample.size()) {
+                        _dim_in = sample.size();
+                        _mapping = Mapping(_dim_in);
+                        reset();
+                    }
+                    if (_dim_out != observation.size()) {
+                        _dim_out = observation.size();
+                        _mean_function = MeanFunction(_dim_out); // the cost of building a functor should be relatively low
+                        reset();
+                    }
+                }
+                else {
+                    assert(sample.size() == _dim_in);
+                    assert(observation.size() == _dim_out);
+                }
+
+                _samples.push_back(sample);
+
+                _observations.conservativeResize(_observations.rows()+1, _dim_out);
+                _observations.bottomRows<1>() = observation.transpose();
+                // _mean_observation = _observations.colwise().mean();
+
+                this->_compute_obs_mean();
+
+                _Phi.conservativeResize(_Phi.rows()+1, _mapping.dim_out());
+                Eigen::VectorXd sample_phi = _mapping.evaluate(sample);
+                _Phi.bottomRows<1>() = sample_phi;
+
+                // this->_compute_incremental_kernel(); 
+
+                // Make Rtilde - [R, sample_phi]
+                Eigen::MatrixXd Rtilde(_dim_mapping+1, _dim_mapping);
```

Comment:
Can't we avoid create a whole new matrix and just re-size `_L`?

Reply:
Do you mean that we should resize `_L` to `(_dim_mapping+1, _dim_mapping)`, do some operations, then resize back to `(_dim_mapping, _dim_mapping)`? Is this better than having a pre-allocated member variable to fill with values at each update?

Reply:
> Do you mean that we should resize _L to (_dim_mapping+1, _dim_mapping), do some operations, then resize back to (_dim_mapping, _dim_mapping)? Is this better than having a pre-allocated member variable to fill with values at each update?

What I mean is something like the following (correct me if I'm wrong and this would not work):

```cpp
// Make Rtilde - [R, sample_phi]
_L = _L.transpose();
_L.conservativeResize(_dim_mapping+1, _dim_mapping);
_L.bottomRows<1>() = sample_phi;

// Note: the Gijsberts paper uses Givens rotations. Is that a bit more efficient?
Eigen::HouseholderQR<Eigen::MatrixXd> qr_res2 = _L.householderQr();
_L = qr_res2.matrixQR().topRows(_dim_mapping).transpose();
```

Reply:
You're right, that should work. I added the change.